Add Euler Flash Loan

Mykyta opened the pull request:
Add Euler on Base, Berachain, and Ethereum chains.
To deploy, use the script
```
forge script script/EulerDeploy.s.sol --broadcast --rpc-url https://base-mainnet.g.alchemy.com/v2/<API_KEY> \
--private-key <PRIVATE_KEY>
```

Mykyta committed:
Add Euler Flash Loan

Mykyta committed:
fix fmt

ultrasecreth commented:
Code wise looks good to me, but the reason we haven't done this is that so far all wrappers are ownerless, to avoid trust assumptions and shit.

I have to admit I also didn't care much cause in Ethereum there're plenty of other sources, but I can see how this could be useful on a more exotic chain.

That said, Euler flashloans are non-reentrant, so if the plan is to use them to lever up on Euler itself, it won't work unless the vault being used for the FL is different than the one borrowing/supplying from/to.

So if that's still good enough, what I'd suggest is to remove the ownership in favour of some permission less method that anybody may call, but with the caveat of checking that the vault is legit and recognised by them.

Alternatively, they have some on-chain registries they call "perspectives" that can be used to get a list of valid vaults, for Ethereum they are:

```
ESCROW_PERSPECTIVES_ADDRESSES: 0xc68CB3658ACf1d49547Fa8605dc158D876cD5828
GOVERNED_PERSPECTIVES_ADDRESSES: 0xC0121817FF224a018840e4D15a864747d36e6Eb2
UNGOVERNED_0X_PERSPECTIVES_ADDRESSES: 0x5345562eD3Ce537582A1A568d3B06c8382Cd60BD
UNGOVERNED_NZX_PERSPECTIVES_ADDRESSES: 0x2206c3341582daAF007BA4569321c8Dd39f0710b
FACTORY_PERSPECTIVES_ADDRESSES: 0xB30f23bc5F93F097B3A699f71B0b1718Fc82e182 - this one is unverified and pulls all in that go through our factory
```

they have a `verifiedArray` method that can be called to list them all, so there could be a permission less method on the wrapper that calls this to iterate and register vaults on a mapping.

So TLDR, imo we must change it to make it permission less in a way that anyone can add more vaults, but in a way where it's somehow verifiable

Mykyta committed:
rework Euler

Mykyta committed:
unused import

Mykyta commented:
Thanks @ultrasecreth appreciate the detailed comment.
I've reworked it by using the`evkFactory`, it significantly decreased the deployment script and potential errors.
Also, as you mentioned, I deleted AccessControl and added the check to avoid 'bad' vaults.

Mykyta committed:
fix fmt

Mykyta commented on script/ERC3156Deploy.s.sol:
```diff
@@ -32,7 +32,7 @@ contract ERC3156Deploy is Script {
         lenders[1] = IERC3156FlashLender(0xb639D208Bcf0589D54FaC24E655C79EC529762B8);
         deployments.push(Deployment(MAINNET, assets, lenders));
-        
+
```
fmt issue

Mykyta commented on src/BaseWrapper.sol:
```diff
@@ -26,6 +26,7 @@ abstract contract BaseWrapper is IERC7399 {
     using SafeERC20 for IERC20;
     using TransientBytes for bytes;
     using TransientBytes for bytes32;
+
```
fmt issue

Mykyta committed:
fix fmt

ultrasecreth commented on src/euler/EulerWrapper.sol:
```diff
@@ -0,0 +1,70 @@
+
+    event VaultsAdded(uint256 count);
+
+    mapping(IERC20 token => IEVault vault) public vaults;
```
you can have multiple vaults for the same asset, I'd make it an array so then you can find a vault that has enough liquidity, and prob sort the array every time we add a new vault for an asset, maybe we can use the IterableSet (or however is called) from OZ so we already remove duplicates? (albeit that'd be unordered, but not sure how much we care about the order as the set is not likely to be big(

Mykyta replied:
Yes, thanks, I didn't think about multiple vaults per token. I've rewritten that logic. I think we can omit a complex data structure to sort vaults by liquidity. As you noticed, it shouldn't contain a lot of vaults per 1 token. Just use an array and iterate through it.

ultrasecreth commented on src/euler/EulerWrapper.sol:
```diff
@@ -0,0 +1,70 @@
+
+    IEVKFactoryPerspective internal factory;
+
+    constructor(address evkFactoryPerspective) {
```
I'd make the constructor type safe already

Mykyta replied:
Yeah, fixed it

Mykyta committed:
add more vaults

ultrasecreth commented on src/euler/EulerWrapper.sol:
```diff
@@ -0,0 +1,105 @@
+                balance = vaultBalance;
+                vault = vaults[i];
+            }
+        }
```
this logic seems duplicated with `maxFlashLoan`, but imo what it should be doing is stopping as soon as it finds a vault with enough liquidity to save gas.

ultrasecreth replied:
we should also check here to see if the vault is not frozen or whatever

Mykyta replied:
Added check to compare requested amount, 16a08839adca32637cce232cb424b0a3a94eeb23

ultrasecreth commented on src/euler/EulerWrapper.sol:
```diff
@@ -0,0 +1,105 @@
+
+        if (tokenVaults[token].length == 0) {
+            tokenVaults[token] = new IEVault[](0);
+        }
```
you don't need to initialise storage arrays

Mykyta replied:
> you don't need to initialise storage arrays

Yes, fixed, 0xC063C3b3625DF5F362F60f35B0bcd98e0fa650fb

ultrasecreth commented on src/euler/EulerWrapper.sol:
```diff
@@ -0,0 +1,105 @@
+        _bridgeToCallback(asset, amount, 0, data);
+
+        // repay amount back to the Euler contract
+        IERC20(asset).transfer(msg.sender, amount);
```
the client is supposed to do this, check the base contract, there's a method to override so when the callback is called, the client is instructed to repay directly, thus saving an ERC20 transfer (can use the uniswap v3 one as an example).

Mykyta replied:
Ok, I see.
I've added that at 45876b5b8e2b178ce9593e8d1a63ee8d595c6dc7

ultrasecreth commented on test/EulerWrapper.t.sol:
```diff
@@ -0,0 +1,101 @@
+        assertEq(wrapper.maxFlashLoan(USDC), 1_985_258_139_987, "Max flash loan not right");
+    }
+
+    function test_flashLoan() external {
```
there's also this vault for USDC `0xC063C3b3625DF5F362F60f35B0bcd98e0fa650fb` that has way less liquidity, so it could be used to test what happens when you request the loan and the first vault in the array is not big enough

Mykyta replied:
Yes, I added this to the test, c0073a93f0935895e500daf8ff47e6f5821570c7

ultrasecreth commented on src/euler/EulerWrapper.sol:
```diff
@@ -0,0 +1,105 @@
+        if (tokenVaults[token].length == 0) {
+            tokenVaults[token] = new IEVault[](0);
+        }
+        tokenVaults[token].push(vault);
```
let's add an event here too :)

Mykyta replied:
This method is called from a constructor in the circle, so it would be annoying to see a lot of events on each added vault

ultrasecreth commented on src/euler/EulerWrapper.sol:
```diff
@@ -0,0 +1,105 @@
+        IEVault[] memory vaults = tokenVaults[IERC20(asset)];
+        uint256 balance = 0;
+
+        for (uint256 i = 0; i < vaults.length; ++i) {
+            uint256 vaultBalance = IERC20(asset).balanceOf(address(vaults[i]));
+            if (vaultBalance > balance) {
+                balance = vaultBalance;
+            }
+        }
```
should we ignore paused vaults?

Mykyta replied:
> should we ignore paused vaults?

I didn't find any useful function from Euler contracts, only [checkVaultStatus](https://github.com/euler-xyz/euler-vault-kit/blob/master/src/EVault/EVault.sol#L156), but I believe it is not what we want to use.

ultrasecreth replied:
yeah, just checked, it's not trivial, and apparently if they get to pause something is cause something is really wrong, so there'll be bigger problems than just flashloans

ultrasecreth commented on src/euler/EulerWrapper.sol:
```diff
@@ -0,0 +1,105 @@
+        emit VaultsAdded(vaultList.length);
+    }
+
+    function addVault(IERC20 token, IEVault vault) public {
```
should a `removeVault` function be added? I wonder if over time we may end up with stale or forever paused vaults. OTOH, that may open the door for someone to "annoy" us by removing random shit, so not sure.

Mykyta replied:
Yeah, good advice, added at 16a08839adca32637cce232cb424b0a3a94eeb23

ultrasecreth reviewed (changes requested):
btw, the "complex" data structure was to avoid duplicates, it doesn't do anything about sorting.

ultrasecreth commented:
@alcueca I'd love a second pair of eyes here :)

Mykyta committed:
refactor

Mykyta committed:
add vault with less liquidity

Mykyta committed:
test liquidity

Mykyta committed:
check duplicated vaults

Mykyta commented:
> btw, the "complex" data structure was to avoid duplicates, it doesn't do anything about sorting.

I added duplicate check on `add` function, cb0444b9cb17ab3af7872f8182d02ab6728bfd95

Mykyta committed:
emit event on removal vault

Mykyta committed:
fix repay

Mykyta committed:
fix repay

Mykyta committed:
fix test

Mykyta commented:
@ultrasecreth thx again for your review, appreciate your help.
Just one question, I didn't find any useful function in Euler Vault, or any other Euler contracts to determine if a vault is freezen/stopped/disable. I checked in this repo https://github.com/euler-xyz/euler-vault-kit/tree/master/src/EVault

ultrasecreth commented:
> @ultrasecreth thx again for your review, appreciate your help. Just one question, I didn't find any useful function in Euler Vault, or any other Euler contracts to determine if a vault is freezen/stopped/disable. I checked in this repo https://github.com/euler-xyz/euler-vault-kit/tree/master/src/EVault

yeah no worries, it's not trivial and unlikely to happen as BAU according to them.

IMO this is ready to merge, but I wonder if you'd like to deploy it wherever you need it and add the address(s) to the README first?

Mykyta commented:
> I wonder if you'd like to deploy it wherever you need it and add the address(s) to the README first?

Yeah, it'd be good to deploy it from some "trusted" address :)
I prefer Berachain as first, others can be Base and Ethereum. See https://github.com/alcueca/erc7399-wrappers/pull/48/files#diff-86a457ccbd123be8a443b1fd13da793b692e11c11e0a3ce898b9d7c96cc9726aR21

Mykyta commented:
@ultrasecreth anyway, can you merge this PR so, so anyone who want it, can deploy it by themselves

Mykyta commented:
@ultrasecreth @alcueca  can you pls merge my PR?
